llvm-runtimes/libgcc/libgcc-19.1.4.ebuild

```
@@ -0,0 +1,47 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit crossdev llvm.org toolchain-funcs
+
+DESCRIPTION="Compiler runtime library for clang, compatible with libgcc_s"
+HOMEPAGE="https://llvm.org/"
+
+LICENSE="Apache-2.0-with-LLVM-exceptions || ( UoI-NCSA MIT )"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~loong ~mips ~ppc64 ~riscv ~x86 ~amd64-linux ~arm64-macos ~ppc-macos ~x64-macos"
+
+DEPEND="
+	llvm-runtimes/compiler-rt:${LLVM_MAJOR}
+	=llvm-runtimes/libunwind-${PV}
+	!!sys-devel/gcc
+"
+BDEPEND="
+	llvm-core/clang:${LLVM_MAJOR}
+"
+
+LLVM_COMPONENTS=( llvm-libgcc )
+llvm.org_set_globals
+
+src_compile() {
+	$(tc-getCC) -E -xc \
+		"${WORKDIR}/llvm-libgcc/gcc_s.ver.in" \
+		-o gcc_s.ver || die
+	$(tc-getCC) -nostdlib \
+		-Wl,--version-script,gcc_s.ver \
+		-Wl,--whole-archive \
+		"${EPREFIX}/usr/lib/clang/${LLVM_MAJOR}/lib/linux/libclang_rt.builtins-${CTARGET%%-*}.a" \
```

```suggestion
		"$(clang --target="${CHOST}" --rtlib=compiler-rt --print-libgcc-file-name)" " \
```

Not sure about `CHOST` vs. `CTARGET`, though.

---

I used `CTARGET` to make sure it works in cross environments.

That said, with your trick, I would just need to make sure that the clang wrappers are used (e.g `aarch64-unknown-linux-musl-clang`).

I think the safest way would be something like:

```
${CTARGET}-clang --rtlib=compiler-rt --print-libgcc-file-name
```

---

You can't run `${CTARGET}` executables on a cross-builder.

---

I'm not sure what do you mean. My intention isn't running any foreign executables. What I want to run is a clang wrapper for the target architecture, to be able to retrieve the runtime file for it. All clang wrappers are just scripts which call clang using a target-specific config file.

Imagine you're on `x86_64-unknown-linux-musl` host, you created an `aarch64-unknown-linux-musl` environment (with `crossdev --llvm --target aarch64-unknown-linux-musl`). Then you emerge libgcc for that target env by doing:

```
aarch64-unknown-linux-musl -av llvm-runtimes/libgcc
```

In that case, `CHOST` is `x86_64-pc-linux-musl` (that's the one usually defined in `make.conf`), `CTARGET` is `aarch64-unknown-linux-musl`. I can execute clang wrappers for both architectures and the output for each of them is:

```
$ x86_64-pc-linux-musl-clang --print-libgcc-file-name
/usr/lib/llvm/19/bin/../../../../lib/clang/19/lib/linux/libclang_rt.builtins-x86_64.a

$ aarch64-unknown-linux-musl-clang --print-libgcc-file-name
/usr/lib/llvm/18/bin/../../../../lib/clang/18/lib/linux/libclang_rt.builtins-aarch64.a
```

Using `CHOST` would break that case, because we would link x86 runtime in a supposedly aarch64 build.

In case of regular native `emerge`, `CHOST` and `CTARGET` are the same (`x86_64-pc-linux-musl`). So calling `${CTARGET}-clang` is still going to work.

---

Right, sorry. Though technically `--target` achieves the same purpose.

---

I went ahead with using clang wrappers (they also specify `--sysroot`), I hope it's fine with you.